lib/services/database/challenge_repository_service.dart

```diff
@@ -0,0 +1,175 @@
+    await _completeChallenge(userDocRef, pid);
+  }
+
+  Future<void> _completeChallenge(
+    DocumentReference parentRef,
+    PostIdFirestore pid,
+  ) async {
+    parentRef
+        .collection(ChallengeFirestore.subCollectionName)
+        .doc(pid.value)
+        .update({
+      ChallengeData.isCompletedField: true,
+    });
+  }
+
+  /// Returns the active challenges of the user with id [uid] that is located at [pos]
+  Future<List<ChallengeFirestore>> getChallenges(
+    UserIdFirestore uid,
+    GeoPoint pos,
+  ) async {
+    final userDocRef =
+        _firestore.collection(UserFirestore.collectionName).doc(uid.value);
+
+    return _getChallenges(userDocRef, pos);
+  }
+
+  Future<List<ChallengeFirestore>> _getChallenges(
+    DocumentReference parentRef,
+    GeoPoint pos,
+  ) async {
+    final challengesCollectionRef =
+        parentRef.collection(ChallengeFirestore.subCollectionName);
+    final pastChallengesCollectionRef = parentRef
+        .collection(ChallengeFirestore.pastChallengesSubCollectionName);
+
+    final challengesSnap = await challengesCollectionRef.get();
+    final pastChallengesSnap = await pastChallengesCollectionRef.get();
+
+    final pastPostIds = pastChallengesSnap.docs
+        .map((doc) => PostIdFirestore(value: doc.id))
+        .toSet();
+
+    final List<ChallengeFirestore> activeChallenges =
+        List.empty(growable: true);
+
+    for (final challengeSnap in challengesSnap.docs) {
+      final challenge = ChallengeFirestore.fromDb(challengeSnap);
+      if (challenge.data.isExpired ||
+          !await _postRepositoryService.postExists(challenge.postId)) {
+        _deleteChallenge(challengeSnap, parentRef);
+        pastPostIds.add(challenge.postId);
+      } else {
+        activeChallenges.add(challenge);
+      }
+    }
+
+    final now = DateTime.now();
+    final sum = now.add(maxChallengeDuration);
+    final expiresOn = DateTime(
+      sum.year,
+      sum.month,
+      sum.day,
+    ); // truncates to the day
+
+    final possiblePosts = await inRangeUnsortedPosts(pos);
+    final postIt = possiblePosts.iterator;
+    final activePostIds = activeChallenges.map((e) => e.postId).toSet();
+    while (activeChallenges.length < maxActiveChallenges && postIt.moveNext()) {
+      final post = postIt.current;
+
+      if (pastPostIds.contains(post) || activePostIds.contains(post)) {
+        continue;
+      }
+
+      final newChallenge = ChallengeFirestore(
+        postId: post,
+        data: ChallengeData(
+          isCompleted: false,
+          expiresOn: Timestamp.fromDate(expiresOn),
+        ),
+      );
+
+      await challengesCollectionRef
+          .doc(post.value)
+          .set(newChallenge.data.toDbData());
+      activeChallenges.add(newChallenge);
+      activePostIds.add(post);
+    }
+
+    return activeChallenges;
+  }
+
+  /// moves the challenge with id [pid] from the active challenges to the past challenges
+  Future<void> _deleteChallenge(
```

Comment:
Good use of `batch()` writes here :+1: 
I wonder if calling the function something like `_moveToPastChallenge` would be more explicit.
What do you think ?